### PR TITLE
perf: replace expensive "Program has OrgUnit" call with faster SQL

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
Deprecated the method `Program.hasOrganisationUnit()` and replaced with SQL method on the `ProgramService`.

All calls to `Program.hasOrganisationUnit()` have been replaced with a call to `ProgramService.hasOrgUnit`, which uses a fast JDBC-based query against the `program_organisationunits` table.

Additionally, added a check to disable the `LOAD` Hibernate listener if ON_LOAD auditing is disabled by the user.

(cherry picked from commit 18a1dead646233bbb969767bb1cd8351be4ae289)

Additional changes in this PR related to the **new Tracker**:

- Removed OrgUnits from Program during object mapping
- Removed OrgUnits list from the Error Code `E1041` output
- Introduce new Preheat class to efficiently preload Program and associated Org Units
- Introduced validation check for Events so that if an Event's OU doesn't belong to the Event's Program OUs, error with code E1029 is raised

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStore.java [hunk 1]
```diff
@@ -105,4 +106,14 @@ public HibernateProgramStore( SessionFactory sessionFactory, JdbcTemplate jdbcTe
...
     }
+
+    public boolean hasOrgUnit( Program program, OrganisationUnit organisationUnit )
+    {
+        NativeQuery<Long> query = getSession().createNativeQuery(
+            "select programid from program_organisationunits where programid = :pid and organisationunitid = :ouid" );
+        query.setParameter( "pid", program.getId() );
+        query.setParameter( "ouid", organisationUnit.getId() );
+
+        return query.getResultList().size() == 1;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStore.java:117]: Maybe slightly safer to do `!query.getResultList().isEmpty()`.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/program/ProgramServiceTest.java [hunk 2]
```diff
@@ -184,4 +184,18 @@ public void testGetProgramByUid()
...
+
+    @Test
+    public void testProgramHasOrgUnit()
+    {
+        programService.addProgram( programA );
+
+        Program p = programService.getProgram( programA.getUid() );
+        OrganisationUnit ou = organisationUnitService.getOrganisationUnit( organisationUnitA.getUid() );
+
+        //sessionFactory.getCurrentSession().flush();
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/program/ProgramServiceTest.java:196]: Remove commented code.

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStore.java (hunk 1 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramStore.java]: Changed.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/program/ProgramServiceTest.java (hunk 2 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/program/ProgramServiceTest.java]: Removed

**Review by stian-sandvold — APPROVED** [review]

**Review by larshelge — APPROVED** [review]

**Review by netroms — APPROVED** [review]

**Pull request merged**